### ISSUE TITLE
Alignment on JSONPath implementations across languages

### ISSUE THREAD

**Issue by cburgmer** [danielaparker/jsoncons#199]
Hi there, I've started comparing all JSONPath implementations I can find to each other: https://cburgmer.github.io/json-path-comparison/.
I'm hoping to start some alignment in the hopes that implementations get a bit more uniform in their interpretation of the standard. I've implemented a lot of adapters for the different languages, I'm just not sure I can get this done for C++. So, contrary to the other libraries I wanted to suggest, that if you are interested to take part in the comparison, maybe you can kick off with an adapter? I would be happy to help. There's some documentation under https://github.com/cburgmer/json-path-comparison#add-an-implementation already.

Part of the project is a set of regression tests, which should allow any library to be tested against the big set of queries already implemented there. Happy also for feedback.

**Comment by danielaparker** [comment]
How would you install the source (include files) from the jsoncons repository? Is support for a particular package manager required?

**Comment by cburgmer** [comment]
Most implementations go via the language's package manager (e.g. https://github.com/cburgmer/json-path-comparison/blob/master/implementations/Ruby_jsonpath/build.ninja), the Bash implementation uses curl (https://github.com/cburgmer/json-path-comparison/blob/master/implementations/Bash_JSONPath.sh/install.sh), the C implementation cheats by using the distribution's (Alpine for docker, locally depends on your OS) way of installing the source (https://github.com/cburgmer/json-path-comparison/blob/master/src/Dockerfile#L25).

**Comment by danielaparker** [comment]
Can you use [vcpkg](https://github.com/microsoft/vcpkg) for install? Currently, this is the only package manager that jsoncons supports.

**Comment by cburgmer** [comment]
Thanks for checking.

It seems there are no binaries available so we would need to install from scratch. In this case that IMHO should be part of the library's build step itself. I wouldn't want to require any developer having to figure out for themselves how to build it. (Something part of the Dockerfile should currently also be installable manually if somebody decided not to go with Docker, for the sake of simplicity.)

I tried installing vcpkg locally on OS X but it seems it's not only incompatible with GCC (fixable) but with OS X 10.14 (`../include/vcpkg/base/files.h:57:41: error: 'file_status' is unavailable: introduced in macOS 10.15`). For me this means I can currently not give you a thumbs up on how easy that would be to use across operating systems.

I can promise to upgrade to OS X 10.15 the next few days, knowing though that there are known breaking changes in that OS version, which makes that a non-trivial one. If you want to wait, I can take care of that the next few days.

**Comment by danielaparker** [comment]
If there's a C++ package manage that you can use, _and_ someone is able to contribute the required files to make jsoncons available through it, I should be able to take responsibility for that going forward.

**Comment by cburgmer** [comment]
I've built a tiny implementation which will fetch the tar from the Github releases page (https://github.com/cburgmer/json-path-comparison/tree/master/implementations/Cpp_jsoncons). Here are two quick findings:

- [ ] `$..*`
  Input:
  ```
  {"key": "value", "another key": {"complex": "string", "primitives": [0, 1]}}
  ```
  Expected output:
  ```
  ["string", "value", 0, 1, [0, 1], {"complex": "string", "primitives": [0, 1]}]
  ```
  Actual output:
  ```
  ["value", {"complex": "string", "primitives": [0, 1]}]
  ```

And for `$['key','another']` many implementations return a list of the two matches in order. Jsoncons however returns the same elements but in reverse order, compare https://cburgmer.github.io/json-path-comparison/results/key_bracket_notation_union.html and https://cburgmer.github.io/json-path-comparison/results/key_bracket_notation_union_with_key_on_nested_objects.html.

**Comment by danielaparker** [comment]
Thanks for the feedback! The two issues reported above should be fixed in the latest release, 0.142.0.
